### PR TITLE
Use Concord interfaces for formatting

### PR DESCRIPTION
**Pull request by cston**

### CODE REVIEW

**Comment by cston** [comment]
@ericfe-ms, @pnelsonmsft, @dotnet/roslyn-interactive

**Comment by pnelsonmsft** [comment]
LGTM - Per offline code review

File: src/ExpressionEvaluator/Core/Source/ResultProvider/ResultProvider.cs [hunk 1]
```diff
@@ -466,20 +576,22 @@ private static string GetTypeName(DkmInspectionContext inspectionContext, DkmClr
             DkmClrCustomTypeInfo declaredTypeInfo,
             DkmInspectionContext inspectionContext,
             string name,
+            string fullName,
+            ReadOnlyCollection<string> formatSpecifiers,
             CompletionRoutine<DkmEvaluationResult> completionRoutine)
         {
-            var type = value.Type.GetLmrType();
-            if (type.IsTypeVariables())
+            var type = value.Type;
```
Comment (KevinH-MS) [inline src/ExpressionEvaluator/Core/Source/ResultProvider/ResultProvider.cs]: Why change this?

File: src/ExpressionEvaluator/Core/Source/ResultProvider/ResultProvider.cs [hunk 2]
```diff
@@ -901,11 +1005,12 @@ internal void Execute()
                     {
                         completionRoutine();
                     }
-                    catch (Exception e)
+                    catch (Exception e) when (ExpressionEvaluatorFatalError.ReportNonFatalException(e, DkmComponentManager.ReportCurrentNonFatalException))
```
Comment (KevinH-MS) [inline src/ExpressionEvaluator/Core/Source/ResultProvider/ResultProvider.cs]: I don't think we want an Exception filter here.  I think we only want the filter in cases where we can't reasonably continue (cases where we'd pass an Exception to the caller).  In this case we're saying it's okay to catch the Exception and continue by calling ```_onException``` (returning a well-formed error results), so I think we should just do that.

We could probably delete ```ExpressionEvaluatorFatalError.ReportNonFatalException```, as I don't see it called elsewhere.

File: src/ExpressionEvaluator/Core/Test/ResultProvider/Debugger/Engine/DkmClrValue.cs [hunk 3]
```diff
@@ -156,7 +151,7 @@ public bool HasUnderlyingString(DkmInspectionContext inspectionContext)
                 throw new ArgumentNullException(nameof(inspectionContext));
             }
 
-            return _formatter.HasUnderlyingString(this, inspectionContext);
+            return inspectionContext.InspectionSession.InvokeFormatter(MethodId.HasUnderlyingString, f => f.HasUnderlyingString(this, inspectionContext));
```
Comment (KevinH-MS) [inline src/ExpressionEvaluator/Core/Test/ResultProvider/Debugger/Engine/DkmClrValue.cs]: Wow...I've never seen this before...  Are we going to add this method to the InspectionContext eventually?

**Comment by KevinH-MS** [comment]
Current iteration looks great!  :+1: 

My only reservation is the Exception filter.

File: src/ExpressionEvaluator/Core/Source/ResultProvider/ResultProvider.cs (hunk 1 above)
Comment (cston) [inline src/ExpressionEvaluator/Core/Source/ResultProvider/ResultProvider.cs]: TypeAndCustomInfo at line 587 requires a DkmClrType now. Reverted this change and passed declaredType to TypeAndCustomInfo instead.

File: src/ExpressionEvaluator/Core/Test/ResultProvider/Debugger/Engine/DkmClrValue.cs (hunk 3 above)
Comment (cston) [inline src/ExpressionEvaluator/Core/Test/ResultProvider/Debugger/Engine/DkmClrValue.cs]: Will ```InvokeFormatter``` be added to ```InspectionContext```? No, this is just the test implementation of a dispatcher.

File: src/ExpressionEvaluator/Core/Source/ResultProvider/ResultProvider.cs (hunk 2 above)
Comment (cston) [inline src/ExpressionEvaluator/Core/Source/ResultProvider/ResultProvider.cs]: Agreed. Removed calls to ```ReportNonFatalException``` and removed method definition.

File: src/ExpressionEvaluator/Core/Test/ResultProvider/Debugger/Engine/DkmClrValue.cs (hunk 3 above)
Comment (KevinH-MS) [inline src/ExpressionEvaluator/Core/Test/ResultProvider/Debugger/Engine/DkmClrValue.cs]: I see.

File: src/ExpressionEvaluator/Core/Source/ResultProvider/Expansion/ArrayExpansion.cs [hunk 4]
```diff
@@ -139,17 +140,14 @@ private static string GetFullName(EvalResultDataItem parent, string name, Format
...
+            var parentRuntimeType = parent.Value.Type;
+            if (!parent.DeclaredTypeAndInfo.Type.Equals(parentRuntimeType.GetLmrType()))
             {
-                bool sawInvalidIdentifier;
-                var parentTypeName = formatter.GetTypeName(new TypeAndCustomInfo(parentRuntimeType), escapeKeywordIdentifiers: true, sawInvalidIdentifier: out sawInvalidIdentifier);
-                if (sawInvalidIdentifier)
+                parentFullName = fullNameProvider.GetClrCastExpression(inspectionContext, parentFullName, parentRuntimeType, customTypeInfo: null, parenthesizeArgument: false, parenthesizeEntireExpression: true);
+                if (parentFullName == null)
                 {
                     return null; // Wouldn't be parseable.
```
Comment (amcasey) [inline src/ExpressionEvaluator/Core/Source/ResultProvider/Expansion/ArrayExpansion.cs]: Consider updating the comment (e.g. saw invalid identifier).

File: src/ExpressionEvaluator/Core/Source/ResultProvider/Expansion/DynamicViewExpansion.cs [hunk 5]
```diff
@@ -78,30 +78,27 @@ private DynamicViewExpansion(DkmClrValue proxyValue, Expansion proxyMembers)
...
             var proxyTypeAndInfo = new TypeAndCustomInfo(_proxyValue.Type);
             var isRootExpression = parent == null;
-            Debug.Assert(isRootExpression != (name == Resources.DynamicView));
             var fullName = isRootExpression ? name : parent.ChildFullNamePrefix;
-            var sawInvalidIdentifier = false;
             var childFullNamePrefix = (fullName == null) ?
                 null :
-                formatter.GetObjectCreationExpression(formatter.GetTypeName(proxyTypeAndInfo, escapeKeywordIdentifiers: true, sawInvalidIdentifier: out sawInvalidIdentifier), fullName);
-            Debug.Assert(!sawInvalidIdentifier); // Expected type name is "Microsoft.CSharp.RuntimeBinder.DynamicMetaObjectProviderDebugView".
+                fullNameProvider.GetClrObjectCreationExpression(inspectionContext, proxyTypeAndInfo.ClrType, proxyTypeAndInfo.Info, fullName);
```
Comment (amcasey) [inline src/ExpressionEvaluator/Core/Source/ResultProvider/Expansion/DynamicViewExpansion.cs]: Comment about why this never returns null?

File: src/ExpressionEvaluator/Core/Source/ResultProvider/ResultProvider.cs [hunk 6]
```diff
@@ -28,31 +27,47 @@ namespace Microsoft.CodeAnalysis.ExpressionEvaluator
...
     {
-        internal readonly Formatter Formatter;
-
         static ResultProvider()
         {
             FatalError.Handler = FailFast.OnFatalException;
         }
 
-        internal ResultProvider(Formatter formatter)
+        internal readonly IDkmClrFormatter2 Formatter2; // Should be removed and replaced with calls through DkmInspectionContext.
```
Comment (amcasey) [inline src/ExpressionEvaluator/Core/Source/ResultProvider/ResultProvider.cs]: Bug number?

File: src/ExpressionEvaluator/Core/Source/ResultProvider/Expansion/ArrayExpansion.cs (hunk 4 above)
Comment (cston) [inline src/ExpressionEvaluator/Core/Source/ResultProvider/Expansion/ArrayExpansion.cs]: Changed.

File: src/ExpressionEvaluator/Core/Source/ResultProvider/Expansion/DynamicViewExpansion.cs (hunk 5 above)
Comment (cston) [inline src/ExpressionEvaluator/Core/Source/ResultProvider/Expansion/DynamicViewExpansion.cs]: If the dynamic view proxy type has an invalid name, GetClrObjectCreationExpression will return null.

File: src/ExpressionEvaluator/Core/Source/ResultProvider/ResultProvider.cs (hunk 6 above)
Comment (cston) [inline src/ExpressionEvaluator/Core/Source/ResultProvider/ResultProvider.cs]: Added reference to https://github.com/dotnet/roslyn/issues/6899.

File: src/ExpressionEvaluator/Core/Source/ResultProvider/ResultProvider.cs [hunk 7]
```diff
@@ -73,6 +91,9 @@ DkmClrValue IDkmClrResultProvider.GetClrValue(DkmSuccessEvaluationResult evaluat
             }
         }
 
+        internal const DkmEvaluationFlags NotRoot = (DkmEvaluationFlags)0x20000;
```
Comment (amcasey) [inline src/ExpressionEvaluator/Core/Source/ResultProvider/ResultProvider.cs]: Are these permanent or are we waiting for new enum members to become available?

File: src/ExpressionEvaluator/Core/Source/ResultProvider/IDkmClrFormatter2.cs [hunk 8]
```diff
@@ -0,0 +1,14 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System.Collections.ObjectModel;
+using Microsoft.VisualStudio.Debugger.Evaluation;
+using Microsoft.VisualStudio.Debugger.Evaluation.ClrCompilation;
+
+namespace Microsoft.VisualStudio.Debugger.ComponentInterfaces
+{
+    public interface IDkmClrFormatter2
```
Comment (amcasey) [inline src/ExpressionEvaluator/Core/Source/ResultProvider/IDkmClrFormatter2.cs]: Should this extend `IDkmClrFormatter`?

File: src/ExpressionEvaluator/Core/Source/ResultProvider/ResultProvider.cs [hunk 9]
```diff
@@ -83,21 +104,78 @@ void IDkmClrResultProvider.GetChildren(DkmEvaluationResult evaluationResult, Dkm
                 return;
             }
 
-            var stackFrame = evaluationResult.StackFrame;
-            GetChildrenAndContinue(dataItem, workList, stackFrame, initialRequestSize, inspectionContext, completionRoutine);
+            var expansion = dataItem.Expansion;
+            if (expansion == null)
+            {
+                var enumContext = DkmEvaluationResultEnumContext.Create(0, evaluationResult.StackFrame, inspectionContext, new EnumContextDataItem(evaluationResult));
+                completionRoutine(new DkmGetChildrenAsyncResult(new DkmEvaluationResult[0], enumContext));
```
Comment (amcasey) [inline src/ExpressionEvaluator/Core/Source/ResultProvider/ResultProvider.cs]: I'm assuming `Array.Empty` is not available here, but maybe we want to introduce our own constant?

File: src/ExpressionEvaluator/Core/Source/ResultProvider/IDkmClrFullNameProvider.cs [hunk 10]
```diff
@@ -0,0 +1,62 @@
...
+    public interface IDkmClrFullNameProvider
+    {
+        /// <summary>
+        /// Return the type name in a form valid in the language (e.g.: escaping keywords)
+        /// or null if the name cannot be represented as valid syntax.
+        /// </summary>
+        string GetClrTypeName(DkmInspectionContext inspectionContext, DkmClrType type, DkmClrCustomTypeInfo customTypeInfo);
+
+        /// <summary>
+        /// Return an array index expression.
```
Comment (amcasey) [inline src/ExpressionEvaluator/Core/Source/ResultProvider/IDkmClrFullNameProvider.cs]: Never returns null, correct?

File: src/ExpressionEvaluator/Core/Source/ResultProvider/IDkmClrFullNameProvider.cs [hunk 11]
```diff
@@ -0,0 +1,62 @@
...
+
+        /// <summary>
+        /// Return a cast expression or null if the type name would be invalid syntax.
+        /// </summary>
+        string GetClrCastExpression(DkmInspectionContext inspectionContext, string argument, DkmClrType type, DkmClrCustomTypeInfo customTypeInfo, bool parenthesizeArgument, bool parenthesizeEntireExpression);
+
+        /// <summary>
+        /// Return an object creation expression or null if the type name would be invalid syntax.
+        /// </summary>
+        /// <remarks>May return null</remarks>
```
Comment (amcasey) [inline src/ExpressionEvaluator/Core/Source/ResultProvider/IDkmClrFullNameProvider.cs]: `<remarks>` not applied consistently.

File: src/ExpressionEvaluator/Core/Source/ResultProvider/IDkmClrFullNameProvider.cs [hunk 12]
```diff
@@ -0,0 +1,62 @@
...
+
+namespace Microsoft.VisualStudio.Debugger.ComponentInterfaces
+{
+    public interface IDkmClrFullNameProvider
+    {
+        /// <summary>
+        /// Return the type name in a form valid in the language (e.g.: escaping keywords)
+        /// or null if the name cannot be represented as valid syntax.
+        /// </summary>
+        string GetClrTypeName(DkmInspectionContext inspectionContext, DkmClrType type, DkmClrCustomTypeInfo customTypeInfo);
```
Comment (amcasey) [inline src/ExpressionEvaluator/Core/Source/ResultProvider/IDkmClrFullNameProvider.cs]: Must call `GetClrValidIdentifier` for all identifiers?

File: src/ExpressionEvaluator/Core/Source/ResultProvider/IDkmClrFullNameProvider.cs [hunk 13]
```diff
@@ -0,0 +1,62 @@
...
+            DkmClrCustomTypeInfo declaringTypeInfo,
+            string memberName,
+            bool memberAccessRequiresExplicitCast,
+            bool memberIsStatic);
+
+        /// <summary>
+        /// Return true if the expression may require parentheses when used
+        /// as a sub-expression in the language.
+        /// </summary>
+        bool ClrExpressionMayRequireParentheses(DkmInspectionContext inspectionContext, string expression);
```
Comment (amcasey) [inline src/ExpressionEvaluator/Core/Source/ResultProvider/IDkmClrFullNameProvider.cs]: I know we like positive names, but I think something like "NeverRequiresParentheses" would be clearer. (Boy, I hope this isn't a name I came up with.)

File: src/ExpressionEvaluator/Core/Source/ResultProvider/IDkmClrFullNameProvider.cs [hunk 14]
```diff
@@ -0,0 +1,62 @@
...
+
+        /// <remarks>May return null</remarks>
+        string GetClrExpressionAndFormatSpecifiers(DkmInspectionContext inspectionContext, string expression, out ReadOnlyCollection<string> formatSpecifiers);
```
Comment (amcasey) [inline src/ExpressionEvaluator/Core/Source/ResultProvider/IDkmClrFullNameProvider.cs]: This feels unrelated to the rest of the interface.

File: src/ExpressionEvaluator/Core/Source/ResultProvider/Formatter.cs [hunk 15]
```diff
@@ -68,16 +173,58 @@ string IDkmClrFormatter.GetUnderlyingString(DkmClrValue value, DkmInspectionCont
...
+                {
+                    case '(':
+                        // Cast, "(A)b", requires parentheses.
+                        if ((parens == 0) && FollowsCloseParen(expr, i))
+                        {
+                            return true;
+                        }
+                        parens++;
+                        break;
+                    case '[':
```
Comment (amcasey) [inline src/ExpressionEvaluator/Core/Source/ResultProvider/Formatter.cs]: Why track square brackets?

File: src/ExpressionEvaluator/Core/Source/ResultProvider/Formatter.cs [hunk 16]
```diff
@@ -68,16 +173,58 @@ string IDkmClrFormatter.GetUnderlyingString(DkmClrValue value, DkmInspectionCont
...
+            int parens = 0;
+            for (int i = 0; i < expr.Length; i++)
             {
-                if (!this.IsIdentifierPartCharacter(ch) && !this.IsWhitespace(ch) && ch != '.') return true;
+                var ch = expr[i];
+                switch (ch)
+                {
+                    case '(':
+                        // Cast, "(A)b", requires parentheses.
+                        if ((parens == 0) && FollowsCloseParen(expr, i))
```
Comment (amcasey) [inline src/ExpressionEvaluator/Core/Source/ResultProvider/Formatter.cs]: From context, we know that parens are always for casts (vs calls)?

File: src/ExpressionEvaluator/CSharp/Test/ResultProvider/FormatSpecifierTests.cs [hunk 17]
```diff
@@ -412,9 +412,9 @@ int H
...
                 children = GetChildren(evalResult, inspectionContext: CreateDkmInspectionContext(DkmEvaluationFlags.ResultsOnly));
                 Verify(children,
-                    EvalResult("F", "{C.<get_F>d__1}", "System.Collections.Generic.IEnumerable<int> {C.<get_F>d__1}", "o.F", DkmEvaluationResultFlags.Expandable | DkmEvaluationResultFlags.ReadOnly),
-                    EvalResult("G", "{C.<get_G>d__3}", "System.Collections.IEnumerable {C.<get_G>d__3}", "o.G", DkmEvaluationResultFlags.Expandable | DkmEvaluationResultFlags.ReadOnly),
-                    EvalResult("H", "3", "int", "o.H", DkmEvaluationResultFlags.ReadOnly));
+                    EvalResult("F", "{C.<get_F>d__1}", "System.Collections.Generic.IEnumerable<int> {C.<get_F>d__1}", "o.F, results", DkmEvaluationResultFlags.Expandable | DkmEvaluationResultFlags.ReadOnly),
+                    EvalResult("G", "{C.<get_G>d__3}", "System.Collections.IEnumerable {C.<get_G>d__3}", "o.G, results", DkmEvaluationResultFlags.Expandable | DkmEvaluationResultFlags.ReadOnly),
+                    EvalFailedResult("H", "Only Enumerable types can have Results View", fullName: null));
```
Comment (amcasey) [inline src/ExpressionEvaluator/CSharp/Test/ResultProvider/FormatSpecifierTests.cs]: Resource string?  Won't this fail in other locales?

Comment (cston) [inline src/ExpressionEvaluator/CSharp/Test/ResultProvider/FormatSpecifierTests.cs]: We're using explicit strings rather than resource ids throughout the Formatter tests.

File: src/ExpressionEvaluator/Core/Source/ResultProvider/Formatter.cs (hunk 16 above)
Comment (cston) [inline src/ExpressionEvaluator/Core/Source/ResultProvider/Formatter.cs]: The parens may be from a call. The string can be an arbitrary C# expression, entered explicitly by the user or generated as a fullname by an EE implementation.

File: src/ExpressionEvaluator/Core/Source/ResultProvider/Formatter.cs (hunk 15 above)
Comment (cston) [inline src/ExpressionEvaluator/Core/Source/ResultProvider/Formatter.cs]: Square brackets are tracked to avoid returning true in all cases where ```NeedsParentheses``` would return true for the expression within brackets (e.g.: ```a[1 + 2]```).

File: src/ExpressionEvaluator/Core/Source/ResultProvider/IDkmClrFormatter2.cs (hunk 8 above)
Comment (cston) [inline src/ExpressionEvaluator/Core/Source/ResultProvider/IDkmClrFormatter2.cs]: Interfaces do not inherit in the debugger API. And inheritance should not be required here in the implementation.

File: src/ExpressionEvaluator/Core/Source/ResultProvider/IDkmClrFullNameProvider.cs (hunk 10 above)
Comment (cston) [inline src/ExpressionEvaluator/Core/Source/ResultProvider/IDkmClrFullNameProvider.cs]: Correct. Updated the comment.

File: src/ExpressionEvaluator/Core/Source/ResultProvider/IDkmClrFullNameProvider.cs (hunk 11 above)
Comment (cston) [inline src/ExpressionEvaluator/Core/Source/ResultProvider/IDkmClrFullNameProvider.cs]: Good catch. Removed.

File: src/ExpressionEvaluator/Core/Source/ResultProvider/IDkmClrFullNameProvider.cs (hunk 14 above)
Comment (cston) [inline src/ExpressionEvaluator/Core/Source/ResultProvider/IDkmClrFullNameProvider.cs]: ```GetClrExpressionAndFormatSpecifiers``` is part of ```IDkmClrFullNameProvider``` since it is used to deconstruct the parent full name when constructing full names for child items. That said, we can remove this method completely if the full name and format specifiers are kept separate in the arguments to ```IDkmClrResultProvider.GetResult```.

File: src/ExpressionEvaluator/Core/Source/ResultProvider/ResultProvider.cs (hunk 7 above)
Comment (cston) [inline src/ExpressionEvaluator/Core/Source/ResultProvider/ResultProvider.cs]: These members should be added to the API.

File: src/ExpressionEvaluator/Core/Source/ResultProvider/IDkmClrFullNameProvider.cs (hunk 13 above)
Comment (cston) [inline src/ExpressionEvaluator/Core/Source/ResultProvider/IDkmClrFullNameProvider.cs]: We should review the entire API.
